### PR TITLE
Support squash skew calculation based on additional namespace size

### PR DESCRIPTION
**Pull request by charleso**
/cc @raronson

### CODE REVIEW

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/squash/SquashReducerLookup.scala [hunk 1]
```diff
@@ -60,10 +72,18 @@ object SquashReducerLookup {
...
-        val count = Math.max(1, (days.toDouble / totalDays * reducers).toInt)
-        lookup.ids.get(fid.toString) -> Int.box(FeatureReducerOffset(i.toShort, count.toShort).toInt)
+        val count = Math.max(1, (days.toDouble / totalDays * reducers(fid.namespace)).toInt)
+        val id = dictionary.byFeatureIndexReverse.getOrElse(fid, 0)
+        Int.box(id) -> Int.box(FeatureReducerOffset(i.toShort, count.toShort).toInt)
     }.toMap.asJava)
   }
+
+  def proportionReducersPerNamespace(partitions: List[Sized[Namespace]], reducers: Int): Map[Namespace, Int] = {
+    val totalSize = partitions.map(_.bytes.toLong).sum
```
Comment (markhibberd) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/squash/SquashReducerLookup.scala]: If you really want to win at life: `paritions.foldMap(_.bytes)`

File: ivory-operation/src/test/scala/com/ambiata/ivory/operation/extraction/squash/SquashReducerLookupSpec.scala [hunk 2]
```diff
@@ -55,4 +57,13 @@ class SquashReducerLookupSpec extends Specification with ScalaCheck { def is = s
...
     SquashWritable.GroupingByFeatureId.hashEntity(bw)
   }
+
+  def proportionAllNamespaces = prop((ns: List[Sized[Namespace]], r: Int) =>
+    SquashReducerLookup.proportionReducersPerNamespace(ns, r).keySet ==== ns.map(_.value).toSet
+  )
+
+  def proportionReducersTotal = prop((s: Sized[Namespace], ns: List[Sized[Namespace]], r: NaturalInt) => {
+    val sns = (s :: ns).groupBy(_.value).mapValues(_.head).values.toList
+    SquashReducerLookup.proportionReducersPerNamespace(sns, r.value).values.sum must beCloseTo(r.value, ns.size)
```
Comment (markhibberd) [inline ivory-operation/src/test/scala/com/ambiata/ivory/operation/extraction/squash/SquashReducerLookupSpec.scala]: Should we make sure it is always `>` as in we don't want any reducer going unused?

**Comment by charleso** [comment]
@markhibberd Thanks for the suggestions/feedback. Does that look OK?

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/squash/SquashReducerLookup.scala [hunk 3]
```diff
@@ -60,10 +73,19 @@ object SquashReducerLookup {
     new ReducerLookup(windowSizes.zipWithIndex.map {
       case ((fid, days), i) =>
         // The number of reducers for this window feature, which is proportional of the number of total window days
-        val count = Math.max(1, (days.toDouble / totalDays * reducers).toInt)
-        lookup.ids.get(fid.toString) -> Int.box(FeatureReducerOffset(i.toShort, count.toShort).toInt)
+        val count = Math.max(1, (days.toDouble / totalDays * reducers(fid.namespace)).toInt)
```
Comment (markhibberd) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/squash/SquashReducerLookup.scala]: As discussed, `totalDays` should now be `totalDays for this namespace only`.

**Comment by charleso** [comment]
@markhibberd Take 2. That may have been a bigger change that was strictly necessarily. Let me know what you think. Take note of how it deals with no-window features now - I _think_ it's better now (previously it wouldn't take up all the reducers).

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/squash/SquashReducerLookup.scala [hunk 4]
```diff
@@ -48,21 +56,71 @@ object SquashReducerLookup {
...
+      f => featuresSizes.getOrElse(f, 1),
+      ns => reducersPerNamespace.getOrElse(ns, 1)
+    )
+  }
+
+  def calculateReducers(windowSizes: Map[FeatureId, Int], totalDays: FeatureId => Int, reducers: Namespace => Int): Map[FeatureId, Int] =
+    // Create a sub-index for all of the window features so they don't overlap on the reducers
+    windowSizes.map {
+      case (fid, days) =>
+        // The number of reducers for this window feature, which is proportional of the number of total window days
```
Comment (markhibberd) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/squash/SquashReducerLookup.scala]: Minor thing, but this comment might need a tweak.

**Comment by markhibberd** [comment]
Looks good I think, I assume you have probably already done this, but did you confirm if you hack https://github.com/ambiata/ivory/pull/611/files#diff-f03489599cdc6dd3cb7a3e9edb9c378dR87 to return the "total size" instead of the "size for this namespace" that it fails those tests.

**Comment by markhibberd** [comment]
In case that wasn't clear: :yellow_heart: 

Tag @raronson for a double check.

**Comment by charleso** [comment]
> "total size" instead of the "size for this namespace" that it fails those tests.

I hadn't, but yah for properties. :)

```
1191347191 is less than 2068768645 (SquashReducerLookupSpec.scala:93)
```

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/squash/SquashReducerLookup.scala [hunk 5]
```diff
@@ -48,21 +56,70 @@ object SquashReducerLookup {
...
+   *   ns2:fid2: 3 months -> 6
+   *   ns2:fid3: 1 months -> 2
+   *
+   *   NOTE: We round-up so in this example ns1 features actually gets 1 more reducer than they are allocated.
+   * }}}
+   */
+  def lookupByNamespaceSizeAndWindow(dictionary: Dictionary, sizes: List[Sized[Namespace]], reducers: Int): Map[FeatureId, Int] = {
+    val windowSizes = calcWindowSizes(dictionary)
+    val featuresSizes = windowSizes.groupBy(_._1.namespace).values.toList.foldMap(x => x.mapValues(_ => x.values.sum))
+    val reducersPerNamespace = proportionReducersPerNamespace(sizes, reducers)
```
Comment (raronson) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/squash/SquashReducerLookup.scala]: types /supernit

File: ivory-operation/src/test/scala/com/ambiata/ivory/operation/extraction/squash/SquashReducerLookupSpec.scala [hunk 6]
```diff
@@ -13,14 +16,19 @@ class SquashReducerLookupSpec extends Specification with ScalaCheck { def is = s
...
+  Proportion namespaces keeps all namespaces                      $proportionAllNamespaces
+  The total of all partition reducers is >= total reducers        $proportionReducersTotal
+  Lookup by window only keeps all features                        $lookupByWindowOnlyAllFeatures
+  The total of all window reducers is >= total reducers           $lookupByWindowOnlyTotal
+  Lookup by namespace size keeps all features                     $lookupByNamespaceSizeAndWindowAllFeatures
+  The total of all sized reducers is >= total reducers            $lookupByNamespaceSizeAndWindowTotal
 """
 
   def lookup = prop((d: VirtualDictionaryWindow, d2: Dictionary, s: Short, e: Int) => {
     val reducers = s & (Short.MaxValue - 1)
```
Comment (raronson) [inline ivory-operation/src/test/scala/com/ambiata/ivory/operation/extraction/squash/SquashReducerLookupSpec.scala]: As discussed, not sure what this does?

**Comment by raronson** [comment]
:+1:  Maybe just some more comments in `SquashReducerLookup` to help follow the code IMO

**Pull request merged**